File: language/compiler/src/com/kubadziworski/parsing/visitor/statement/BlockStatementVisitor.java
package com.kubadziworski.parsing.visitor.statement;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.BlockContext;
import jiva.parser.JivaParser.StatementContext;
import com.kubadziworski.domain.scope.Scope;
import com.kubadziworski.domain.node.statement.Block;
import com.kubadziworski.domain.node.statement.Statement;
import org.antlr.v4.runtime.misc.NotNull;

import java.util.List;
import java.util.stream.Collectors;

public class BlockStatementVisitor extends JivaBaseVisitor<Block>{
    private final Scope scope;

    public BlockStatementVisitor(Scope scope) {
        this.scope = scope;
    }

    @Override
    public Block visitBlock(@NotNull BlockContext ctx) {
        List<StatementContext> blockstatementsCtx = ctx.statement();
        Scope newScope = new Scope(scope);
        StatementVisitor statementVisitor = new StatementVisitor(newScope);
        List<Statement> statements = blockstatementsCtx.stream().map(smtt -> smtt.accept(statementVisitor)).collect(Collectors.toList());
        return new Block(newScope, statements);
    }
}

File: language/compiler/src/parsing/visitor/statement/BlockStatementVisitor.kt
package jiva.parsing.visitor.statement

import jiva.domain.node.statement.Block
import jiva.domain.scope.Scope
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.BlockContext

class BlockStatementVisitor(private val scope: Scope) : JivaBaseVisitor<Block>() {

    override fun visitBlock(ctx: BlockContext): Block {
        val blockStatementsCtx = ctx.statement()
        val newScope = Scope(scope)
        val statementVisitor = StatementVisitor(newScope)
        val statements = blockStatementsCtx.map { it.accept(statementVisitor)!! }
        return Block(newScope, statements)
    }

}
